% Template for PLoS
% Version 3.4 January 2017
%
% % % % % % % % % % % % % % % % % % % % % %
%
% -- IMPORTANT NOTE
%
% This template contains comments intended 
% to minimize problems and delays during our production 
% process. Please follow the template instructions
% whenever possible.
%
% % % % % % % % % % % % % % % % % % % % % % % 
%
% Once your paper is accepted for publication, 
% PLEASE REMOVE ALL TRACKED CHANGES in this file 
% and leave only the final text of your manuscript. 
% PLOS recommends the use of latexdiff to track changes during review, as this will help to maintain a clean tex file.
% Visit https://www.ctan.org/pkg/latexdiff?lang=en for info or contact us at latex@plos.org.
%
%
% There are no restrictions on package use within the LaTeX files except that 
% no packages listed in the template may be deleted.
%
% Please do not include colors or graphics in the text.
%
% The manuscript LaTeX source should be contained within a single file (do not use \input, \externaldocument, or similar commands).
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% -- FIGURES AND TABLES
%
% Please include tables/figure captions directly after the paragraph where they are first cited in the text.
%
% DO NOT INCLUDE GRAPHICS IN YOUR MANUSCRIPT
% - Figures should be uploaded separately from your manuscript file. 
% - Figures generated using LaTeX should be extracted and removed from the PDF before submission. 
% - Figures containing multiple panels/subfigures must be combined into one image file before submission.
% For figure citations, please use "Fig" instead of "Figure".
% See http://journals.plos.org/plosone/s/figures for PLOS figure guidelines.
%
% Tables should be cell-based and may not contain:
% - spacing/line breaks within cells to alter layout or alignment
% - do not nest tabular environments (no tabular environments within tabular environments)
% - no graphics or colored text (cell background color/shading OK)
% See http://journals.plos.org/plosone/s/tables for table guidelines.
%
% For tables that exceed the width of the text column, use the adjustwidth environment as illustrated in the example table in text below.
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% -- EQUATIONS, MATH SYMBOLS, SUBSCRIPTS, AND SUPERSCRIPTS
%
% IMPORTANT
% Below are a few tips to help format your equations and other special characters according to our specifications. For more tips to help reduce the possibility of formatting errors during conversion, please see our LaTeX guidelines at http://journals.plos.org/plosone/s/latex
%
% For inline equations, please be sure to include all portions of an equation in the math environment.  For example, x$^2$ is incorrect; this should be formatted as $x^2$ (or $\mathrm{x}^2$ if the romanized font is desired).
%
% Do not include text that is not math in the math environment. For example, CO2 should be written as CO\textsubscript{2} instead of CO$_2$.
%
% Please add line breaks to long display equations when possible in order to fit size of the column. 
%
% For inline equations, please do not include punctuation (commas, etc) within the math environment unless this is part of the equation.
%
% When adding superscript or subscripts outside of brackets/braces, please group using {}.  For example, change "[U(D,E,\gamma)]^2" to "{[U(D,E,\gamma)]}^2". 
%
% Do not use \cal for caligraphic font.  Instead, use \mathcal{}
%
% % % % % % % % % % % % % % % % % % % % % % % % 
%
% Please contact latex@plos.org with any questions.
%
% % % % % % % % % % % % % % % % % % % % % % % %

\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}


% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\setlength{\headheight}{27.023pt}
\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\sf PLOS}

%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}

%% END MACROS SECTION


\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{Trimming DNA Reads} % Please use "sentence case" for title and headings (capitalize only the first word in a title (or heading), the first word in a subtitle (or subheading), and any proper nouns).
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Arvin Chireh\textsuperscript{1,*}
\\
\bigskip
\textbf{1} Institutionen f√∂r Klinisk Neurovetenskap, Karolinska Institutet, Stockholm, Sweden
\\

\bigskip

% Insert additional author notes using the symbols described below. Insert symbol callouts after author names as necessary.
% 
% Remove or comment out the author notes below if they aren't used.
%
% Primary Equal Contribution Note
%\Yinyang These authors contributed equally to this work.

% Additional Equal Contribution Note
% Also use this double-dagger symbol for special authorship notes, such as senior authorship.
%\ddag These authors also contributed equally to this work.

% Current address notes
%\textcurrency Current Address: Dept/Program/Center, Institution Name, City, State, Country % change symbol to "\textcurrency a" if more than one current address note
% \textcurrency b Insert second current address 
% \textcurrency c Insert third current address

% Deceased author note
%\dag Deceased

% Group/Consortium Author Note
%\textpilcrow Membership list can be found in the Acknowledgments section.

% Use the asterisk to denote corresponding authorship and provide email address in note below.
* arvin.chireh@gmail.com

\end{flushleft}
% Please keep the abstract below 300 words
\section*{Abstract}
In modern DNA-sequencing, the quality of the base calls frequently drops towards the end of the sequencing reads. Therefore, there is a need for programs that can recognize this technical quality decline, and trim the ends of reads accordingly. The purpose of this project was to develop and test a new program for trimming of DNA sequences.

A program with two alternative trimming algorithms (tr1 and tr2) was written in Python. A fastq file produced with MISEQ sequencer, containing 250 000 reads, was used for evaluating the program. Mean sequence length and base call quality was assessed before and after the trimming, with corresponding histograms.

The tr1 algorithm trimmed 104 960 of  250 000 reads (42 \%), reducing the mean sequence length from 169 to 149, increasing mean phred score from 29.4 to 29.9. The tr2 algorithm trimmed 34877 of 250 000 reads (14\%), reducing mean sequence length from 169 to 157, improving mean phred score from 29.4 to 30. Only descriptive statistics was used.

In conclusion, the trimming program developed and tested here gives a marginal increase in mean base call quality, at the cost of reduced sequencing length. The algorithms were only tested on one dataset. Further evaluation is needed to properly benchmark the algorithms. Most likely, one should look for other, more efficient and more selective algorithms.

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step. 
% Author Summary not valid for PLOS ONE submissions.   
% \section*{Author summary}


\linenumbers

% Use "Eq" instead of "Equation" for equation citations.
\section*{Introduction}
Recent advances in DNA sequencing has made it possible to sequence DNA at an unprecedented pace. As a result, DNA sequencing has emerged as a valuable tool in biological research as well as clinical decision making. However, the current techniques have some limitations. For instance, base call quality frequently drops towards the end of each individual read from the sequencer. Therefore, there is a need for software than can recognize this quality decline towards the end of reads, and remove (trim) low quality bases. Quality, in this context, is described in terms of a phred score \cite{{bib1}} , ranging between 0 and 40. The purpose of this work was to develop and evaluate a new method of trimming low quality bases without unnecessary removal of good quality bases.



\section*{Materials and methods}
\subsection*{Test data}
A fastq file generated from the MISEQ instrument (llumina, Inc), containing 250 000 DNA sequencing reads, was used as test data for assessing the effect of the algorithm. An empty file, a fasta file and a fastq file with short sequences (15 base pair sequences) was used to check for the ability of the program to cope with suboptimal data.

\subsection*{Program}
The program was written as an executable python script. Both algorithms, presented below, were written in the same script, giving the user the opportunity to choose algorithm when calling the script.

Several controls were built into the program. One control was written to check for empty files or wrong file types. Another control was inserted to skip trimming of too short sequences. No control for checking nucelotide validity (i.e. allowed base) was implemented, as it would slow down the program, and was not considered relevant for the purpose.

The "tr1" algorithm was designed to calculate the mean phred score and its standard deviation (SD) of the first 30 bases in each record, to use as an indicator. For each record in the fastq file, the program iterates through each base (base[i]) and checks if the mean phred score from i and onwards (seq[i:]) is below the quality threshold that was set (1 SD or more below the mean phred score of the first 30 bases). If the condition is met, the sequence is trimmed at that point and stored in a list of trimmed records. Sequences that are too short (less than 30 bases), or fail to fulfill criteria for trimming, are stored in the list of trimmed records as is. The program reports number of trimmed, not tirmmed (denoted as "untrimmed") and too short sequences.

The "tr2" algorithm is based on a similar approach. However, instead of checking the mean phred score from base[i]  and onwards, the program looks at 10 consecutive bases from i and onwards (i.e. base[i], base[i+1] ... base[i+9]). If all these consecutive bases are below the quality threshold, the sequence is cut at base[i].

The program writes the new records to a fastq file.

\subsection*{Assessment}
Mean sequencing length and mean quality (phred score), together with equivalent metrics for the last 20 bases in each record (i.e. the tails of the sequences) were assessed before and after trimming. Histograms were plotted to check the change of sequencing length and quality distributions.



% Results and Discussion can be combined.
\section*{Results}
The tr1 algorithm trimmed 104 960 of  250 000 reads (42 \%), reducing the mean sequence length from 169 to 149, increasing mean phred score from 29.4 to 29.9, Table~\ref{table1}, Figure~\ref{fig1} . The tr2 algorithm trimmed 34877 of 250 000 reads (14\%), reducing mean sequence length from 169 to 157, improving mean phred score from 29.4 to 30, Table~\ref{table2}, Figure~\ref{fig2}.

\begin{table}[!ht]
%\begin{adjustwidth}{-2.25in}{0in} % Comment out/remove adjustwidth environment if table fits in text column.
\centering
\caption{
{\bf Statistics for tr1 algorithm} Tail defined as the last 20 bases.}

\begin{tabular}{|l|l|l|l|}
\hline
                & Seq length, mean & phred score, mean & tail phred score, mean \\ \hline
Before trimming & 169 &        29.4    &    28.6    \\ \hline
After trimming  &   149               &        29.9    &    30.4    \\ \hline
\end{tabular}
\label{table1}
\end{table}

\begin{table}[!ht]
%\begin{adjustwidth}{-2.25in}{0in} % Comment out/remove adjustwidth environment if table fits in text column.
\centering
\caption{
{\bf Statistics for tr2 algorithm} Tail defined as the last 20 bases.}

\begin{tabular}{|l|l|l|l|}
\hline
                & Seq length, mean & phred score, mean & tail phred score, mean \\ \hline
Before trimming & 169 &        29.4    & 28.6       \\ \hline
After trimming  &   157               &        30   & 29.1        \\ \hline
\end{tabular}
\label{table2}
\end{table}

% Place figure captions after the first paragraph in which they are cited.
\begin{figure}[!h]
\caption{{\bf Tr1 Algorithm}
Histograms showing distributions of sequencing length and quality, for algorithm tr1. Untrimmed = Before trimming. Trimmed = After trimming.}
\includegraphics[width=\linewidth]{../../results/trimming1/miseq2/seqlength.pdf}
\includegraphics[width=\linewidth]{../../results/trimming1/miseq2/seqquality.pdf}
\label{fig1}
\end{figure}

\begin{figure}[!h]
\caption{{\bf Tr2 Algorithm}
Histograms showing distributions of sequencing length and quality, for algorithm tr2. Untrimmed = Before trimming. Trimmed = After trimming.}
\includegraphics[width=\linewidth]{../../results/trimming2/miseq2/seqlength.pdf}
\includegraphics[width=\linewidth]{../../results/trimming2/miseq2/seqquality.pdf}
\label{fig2}
\end{figure}


\section*{Discussion}
In this work, I attempted to develop a software to trim low quality reads at the end of sequencing reads. The developed program contains two algorithms, however neither of these seem to do major improvements in overall quality. However, the expected quality improvement is highly dependent on the sequencing data. In our test data, the mean phred quality of the last 20 bases (i.e. tail) was not very different from the overall base call quality (29.4 mean for whole sequences, 28.6 mean for tails). As a result, one should not expect high changes in overall quality following a trimming of the tail. In fact, both algorithms succeeded in making the mean quality of the tail close to the overall mean, or even better, particularly tr1. As a disadvantage of the program design, all reported quality statistics are means of all bases pooled together. There was no pairwise comparison for each sequence. 

More worrisome, only 42 \% and 14 \% of the reads were trimmed by tr1 and tr2, respectively. In other words, the algorithm failed to recognize and trim low quality bases in the tails of a majority of the sequences. This may or may not be dependent on the data set.

\subsection*{Notes on management of the project}
This project was also an exercise with the purpose of improving the ability to handle a bioinformatics project. As a result, I read and implemented some of the ideas introduced by William Stafford Noble \cite{bib2}. For instance, I used a modified version of the proposed directory structure, with logical and chronological sub-directories. Moreover, I kept a digital lab notebook in markdown format, with procedural notes coupled with timestamps. I also used git version control system to keep versions of my software, and looked at previous version at some points during the development process.

Noble suggest keeping a file with all command-lines that were used, and keep this file in parallel to the logbook. I personally find it easier to just keep those commands in the logbook as well, together with comments and other text. I find driver scripts too complicated at this point, and maybe not necessary for the scale of projects at this level.

I also added control code in my program to detect some primitive errors, as suggested by Noble and the assignment instructions. Obviously, these controls could have been more carefully designed, and cover a larger range of  potential problems.

\section*{Conclusion}
In conclusion, the algorithms used in my program offers a modest quality improvement at the cost of sequence data. I do recommend a search for more effective and selective algorithms. Moreover, the algortihms has to be tested on a further range of data sets before defintive conclusions can be made, perhaps with pairwise testing and/or other statistical testing.

This project was a great opportunity to improve management of a bioinformatic project. The ideas of William Stafford Noble were considered and to a large degree implemented in this work.



\nolinenumbers

% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% or
%
% Compile your BiBTeX database using our plos2015.bst
% style file and paste the contents of your .bbl file
% here. See http://journals.plos.org/plosone/s/latex for 
% step-by-step instructions.
% 
\begin{thebibliography}{10}
\bibitem{bib1}
Ewing B, Green P.
\newblock {Base-calling of automated sequencer traces using phred. II. Error probabilities.
}.
\newblock Genome Res. 1998 Mar;8(3):186-94.

\bibitem{bib2}
William Stafford Noble.
\newblock {A Quick Guide to Organizing Computational Biology Projects}.
\newblock PLoS Comput Biol. 2009 Jul; 5(7): e1000424.



\end{thebibliography}



\end{document}

